File: OBJC/Tripcode.m
//
//  Tripcode.m
//  Tripcode iOS Objective-C
//
//  Created by pat on 12/13/17.
//  Copyright © 2017 Micromat, Inc. All rights reserved.
//

#import "Tripcode.h"

@implementation Tripcode

static const char * s_tripcodeDigits = "23456789BCDFGHJKLMNPQRSTVWXZ";

#define TC_NBR_DIGITS        10
#define TC_MAX_DEGREE        360

+ (NSString *)tripcodeForLocation:(CLLocation *)location
{
    return [self tripcodeForCoordinate: location.coordinate];
}

+ (NSString *)tripcodeForCoordinate:(CLLocationCoordinate2D)coordinate
{
    long baseTable[TC_NBR_DIGITS / 2];
    double accuracy = [self initValuesFor: baseTable
                                  reverse: NULL];
    
    double latitude = coordinate.latitude;
    double longitude = coordinate.longitude;
    
    if (latitude < -90.0 || latitude > 90.0 || longitude < -180.0 || longitude > 180.0)
        return nil;
    
    long llatitude = round((latitude + 90.0) / (accuracy / 2.0));
    long llongitude = round((longitude + 180.0 ) / accuracy);
    
    // Encode the latitude and longitude using the base table and tripcode digits
    NSMutableString * tripcode = [NSMutableString new];
    
    int i;
    for (i = (TC_NBR_DIGITS / 2) - 1; i >= 0; i --)
    {
        [tripcode appendFormat: @"%c", s_tripcodeDigits[llatitude / baseTable[i]]];
        llatitude %= baseTable[i];
        
        if (tripcode.length == 3)
            [tripcode appendString: @"-"];
        
        [tripcode appendFormat: @"%c", s_tripcodeDigits[llongitude / baseTable[i]]];
        llongitude %= baseTable[i];
        
        if (tripcode.length == 7)
            [tripcode appendString: @"-"];
    }
    
    return tripcode;
}

+ (CLLocation *)locationForTripcode:(NSString *)tripcode
{
    return [[CLLocation alloc] initWithCoordinate: [self coordinateForTripcode: tripcode]
                                         altitude: 0.0
                               horizontalAccuracy: 0.0
                                 verticalAccuracy: 0.0
                                        timestamp: [NSDate date]];
}

+ (double)initValuesFor:(long *)baseTable reverse:(char *)reverseTable
{
    // Initialize the base table
    unsigned long baseNbr = strlen(s_tripcodeDigits);
    unsigned long i, factor;
    
    for (i = 0, factor = 1; i < (TC_NBR_DIGITS / 2); i ++, factor *= baseNbr)
    {
        if (baseTable)
            baseTable[i] = factor;
    }
    
    // Determine the accuracy
    double accuracy = ((double)TC_MAX_DEGREE) / ((double)factor - 1.0);
    
    // We need to truncate the accuracy to minimize disparity between various system and potential rouding error
    factor = (long)pow(10.0, (int)(-log10(accuracy)) + 2);
    accuracy = ceil(accuracy * (double)factor) / (double)factor;
    
    if (!reverseTable)
        return accuracy;
    
    // Initialize the reverse digits table
    memset(reverseTable, -1, 1);
    for (i = 0; i < strlen(s_tripcodeDigits); i ++)
    {
        char digit = s_tripcodeDigits[i];
        reverseTable[digit] = i;
        if (digit >= 'A' && digit <= 'Z')
            reverseTable[digit + 0x20] = i;    // support for lower case
    }
    
    return accuracy;
}

+ (CLLocationCoordinate2D)coordinateForTripcode:(NSString *)tripcode
{
    long baseTable[(TC_NBR_DIGITS / 2)];
    char reverseTable[127];
    double accuracy = [self initValuesFor: baseTable
                                  reverse: reverseTable];

    long llatitude = 0;
    long llongitude = 0;
    long digit;
    
    char tripcodeX[TC_NBR_DIGITS * 2];
    char tc8extra[3] = {s_tripcodeDigits[0], s_tripcodeDigits[0], 0};
    strcpy (tripcodeX, tripcode.UTF8String);
    strcat (tripcodeX, tc8extra);
    
    CLLocationCoordinate2D invalidTripcode = {999.0, 999.0};
    
    for (int i = (TC_NBR_DIGITS / 2) - 1, j = 0; i >= 0 && tripcodeX[j];)
    {
        while ((digit = tripcodeX[j]) == '-')
            j++;
        
        if (digit >= 127)
            return invalidTripcode;
        
        digit = reverseTable[digit];
        if (digit == -1)
            return invalidTripcode;
        
        llatitude += digit * baseTable[i];
        j ++;
        
        while ((digit = tripcodeX[j]) == '-')
            j++;
        if (digit >= 127)
            return invalidTripcode;
        
        digit = reverseTable[digit];
        if (digit == -1)
            return invalidTripcode;
        
        llongitude += digit * baseTable[i];
        j ++;
        i --;
    }
    
    double latitude = ((double)llatitude * (accuracy / 2.0)) - 90.0;
    double longitude = ((double)llongitude * accuracy) - 180.0;
    
    return (CLLocationCoordinate2D){latitude, longitude};
}

@end

File: PHP/tripcode.php
<?php

/*
 *  tripcode.php
 *  tripcode
 *
 *  Created by Jean-François Ducarroz on 1/8/11.
 *  Copyright 2011 Outspring, Inc. All rights reserved.
 *
 */

class Tripcode
{
    private $m_tripcodeDigits = NULL;
    private $m_baseTable = array();
    private $m_reverseTable = array();
    private $m_accuracy = 0;

    function __construct()
    {
        define('TC_NBR_DIGITS', 10);
        define('TC_MAX_DEGREE', 360);

        // Convert the tripcode digits into an array
        $this->m_tripcodeDigits = preg_split('//', "23456789BCDFGHJKLMNPQRSTVWXZ", -1, PREG_SPLIT_NO_EMPTY);

        // Initialize the base table
        $baseNbr = count($this->m_tripcodeDigits);

        for ($i = 0, $factor = 1; $i < (int)((TC_NBR_DIGITS + 1) / 2); $i ++, $factor *= $baseNbr)
            $this->m_baseTable[$i] = $factor;

        // Determine the accuracy
        $this->m_accuracy = ((double)TC_MAX_DEGREE) / ((double)$factor - 1.0);
        
        // We need to truncate the accuracy to minimize disparity between various system and potential rouding error
        $factor = (int)pow(10.0, (int)(-log10($this->m_accuracy)) + 2);
        $this->m_accuracy = ceil($this->m_accuracy * (double)$factor) / (double)$factor;
        
        // Initialize the reverse digits table
        $this->m_reverseTable = array_flip($this->m_tripcodeDigits);
    }

    public function tripcodeForCoordinate($latitude, $longitude)
    {
        if ($latitude < -90.0 || $latitude > 90.0 || $longitude < -180.0 || $longitude > 180.0)
            return NULL;

        $llatitude = round(($latitude + 90.0) / ($this->m_accuracy / 2.0));
        $llongitude = round(($longitude + 180.0 ) / $this->m_accuracy);
        
        // Encode the latitude and longitude using the base table and tripcode digits
        $tripcode = '';
        for ($i = (int)((TC_NBR_DIGITS + 1) / 2) - 1, $j = 0; $i >= 0; $i --)
        {
            $tripcode .= $this->m_tripcodeDigits[$llatitude / $this->m_baseTable[$i]];
            $llatitude %= $this->m_baseTable[$i];
            
            if (++ $j == 3)
            {
                $tripcode .= '-';
                $j ++;
            }

            $tripcode .= $this->m_tripcodeDigits[$llongitude / $this->m_baseTable[$i]];
            $llongitude %= $this->m_baseTable[$i];
            
            if (++ $j == 7)
            {
                $tripcode .= '-';
                $j ++;
            }
        }

        return $tripcode;
    }
    
    public function coordinateForTripcode($tripcode)
    {
        $location = array();
        
        $llatitude = 0;
        $llongitude = 0;
        $digit;
        
        $tripcodeX = $tripcode.$this->m_tripcodeDigits[0].$this->m_tripcodeDigits[0];
        $len = strlen($tripcodeX);

        for ($i = (int)((TC_NBR_DIGITS + 1) / 2) - 1, $j = 0; $i >= 0 && $j < $len;)
        {
            while (($digit = $tripcodeX[$j]) == '-')
                $j++;
            if (ord($digit) >= 127)
                return $location;
            
            if (isset($this->m_reverseTable[$digit]))
                $digit = $this->m_reverseTable[$digit];
            else
                return $location;
            
            $llatitude += $digit * $this->m_baseTable[$i];
            $j ++;
            
            while (($digit = $tripcodeX[$j]) == '-')
                $j++;
            if (ord($digit) >= 127)
                return $location;
            
            if (isset($this->m_reverseTable[$digit]))
                $digit = $this->m_reverseTable[$digit];
            else
                return $location;

            $llongitude += $digit * $this->m_baseTable[$i];
            $j ++;

            $i --;
        }

        $location['latitude'] = ((double)$llatitude * ($this->m_accuracy / 2.0)) - 90.0;
        $location['longitude'] = ((double)$llongitude * $this->m_accuracy) - 180.0;

        return $location;
    }
    
    public function tripcodeWithOffset($tripcode, $offset)
    {
        $location = array();
        
        $llatitude = 0;
        $llongitude = 0;
        $digit;
        
        $tripcodeX = $tripcode.$this->m_tripcodeDigits[0].$this->m_tripcodeDigits[0];
        $len = strlen($tripcodeX);

        for ($i = (int)((TC_NBR_DIGITS + 1) / 2) - 1, $j = 0; $i >= 0 && $j < $len;)
        {
            while (($digit = $tripcodeX[$j]) == '-')
                $j++;
            if (ord($digit) >= 127)
                return $location;
            
            if (isset($this->m_reverseTable[$digit]))
                $digit = $this->m_reverseTable[$digit];
            else
                return $location;
            
            $llatitude += $digit * $this->m_baseTable[$i];
            $j ++;
            
            while (($digit = $tripcodeX[$j]) == '-')
                $j++;
            if (ord($digit) >= 127)
                return $location;
            
            if (isset($this->m_reverseTable[$digit]))
                $digit = $this->m_reverseTable[$digit];
            else
                return $location;

            $llongitude += $digit * $this->m_baseTable[$i];
            $j ++;

            $i --;
        }
        
        $llatitude += $offset;
        $llongitude += $offset;
        
        // Encode the latitude and longitude using the base table and tripcode digits
        $tripcode = '';
        for ($i = (int)((TC_NBR_DIGITS + 1) / 2) - 1, $j = 0; $i >= 0; $i --)
        {
            $tripcode .= $this->m_tripcodeDigits[$llatitude / $this->m_baseTable[$i]];
            $llatitude %= $this->m_baseTable[$i];
            
            if (++ $j == 3)
            {
                $tripcode .= '-';
                $j ++;
            }

            $tripcode .= $this->m_tripcodeDigits[$llongitude / $this->m_baseTable[$i]];
            $llongitude %= $this->m_baseTable[$i];
            
            if (++ $j == 7)
            {
                $tripcode .= '-';
                $j ++;
            }
        }

        return $tripcode;
    }
}

?>
